R/sagemaker.R

```diff
@@ -0,0 +1,481 @@
+#' @param enable_network_isolation A logical to specify whether the container
+#' will run in network isolation mode. Defaults to `FALSE`.
+#' @param tags A list of tags for labeling the Amazon SageMaker model or
+#' model endpint to be created.
+#' @rdname vetiver_sm_build
+#' @export
+vetiver_sm_model <- function(image_uri,
+                             model_name,
+                             role = NULL,
+                             vpc_config = list(),
+                             enable_network_isolation = FALSE,
+                             tags = list()) {
+    check_installed(c("smdocker", "paws.machine.learning"))
+    config <- smdocker::smdocker_config()
+    sagemaker_client <- paws.machine.learning::sagemaker(config)
+
+    if (is_missing(model_name)) {
+        model_name <- base_name_from_image(image_uri)
+    }
+
+    if (is.null(role)) {
+        role <- smdocker::sagemaker_get_execution_role()
+    }
+
+    request <- list(
+        "ModelName" = model_name,
+        "ExecutionRoleArn" = role,
+        "PrimaryContainer" = list("Image" = image_uri)
+    )
+    request$Tags <- .append_project_tags(tags)
+    request$VpcConfig <- check_vpc_config(vpc_config)
+    if (isTRUE(enable_network_isolation)) {
+        request$EnableNetworkIsolation <- TRUE
+    }
+    # create model
+    do.call(sagemaker_client$create_model, request)
+
+    return(model_name)
+}
+
+#' @param endpoint_name The name to use for the Amazon SageMaker model endpoint
+#' to be created, if to be different from `model_name`.
+#' @param instance_type Type of EC2 instance to use; see
+#' [Amazon SageMaker pricing](https://aws.amazon.com/sagemaker/pricing/).
+#' Defaults to `"ml.t2.medium"`.
+#' @param initial_instance_count The initial number of instances to run
+#' in the endpoint.
+#' @param accelerator_type Type of Elastic Inference accelerator to
+#' attach to an endpoint for model loading and inference, for
+#' example, `"ml.eia1.medium"`.
+#' @param tags A list of tags for labeling the Amazon SageMaker model endpoint.
+#' @param kms_key The ARN of the KMS key used to encrypt the data on the
+#' storage volume attached to the instance hosting the endpoint.
+#' @param data_capture_config A list for configuration to control how Amazon
+#' SageMaker captures inference data.
+#' @param volume_size The size, in GB, of the ML storage volume attached to
+#' the individual inference instance associated with the production variant.
+#' Currently only Amazon EBS gp2 storage volumes are supported.
+#' @param model_data_download_timeout The timeout value, in seconds, to download
+#' and extract model data from Amazon S3.
+#' @param wait A logical for whether to wait for the endpoint to be deployed.
+#' Defaults to `TRUE`.
+#' @rdname vetiver_sm_build
+#' @export
+vetiver_sm_endpoint <- function(model_name,
+                                endpoint_name = NULL,
+                                instance_type = "ml.t2.medium",
```

Comment:
I added a non-`NULL` default value here, because the instance type is not optional. This filters up to the main `vetiver_deploy_sagemaker()` function as well. I think we have two options here:

- set a default like this (thoughts on whether this is the right one?)
- require the user to put in a value (no default) in which case it should go with the other required arguments in `vetiver_deploy_sagemaker()`

The second option would make the function signature look something like this, which does strike me as weird:

```r
vetiver_deploy_sagemaker <- function(board,
                                     name,
                                     instance_type,
                                     version = NULL,
                                     predict_args = list(),
                                     docker_args = list(),
                                     repo_name = glue("vetiver-sagemaker-{name}"),
                                     compute_type = c(
                                         "BUILD_GENERAL1_SMALL",
                                         "BUILD_GENERAL1_MEDIUM",
                                         "BUILD_GENERAL1_LARGE",
                                         "BUILD_GENERAL1_2XLARGE"
                                     ),
                                     ...) {
<<CODE>>
}
```

I don't like how the pin arguments are getting mixed up with others. What do you think @DyfanJones?

Reply:
Sorry I am not a 100% on what the 2 options are sorry :s

From what I understood:

- Is having `instance_type` with a default"ml.t2.medium" ok?
    - I am not sure, I don't think `ml.t2.medium` isn't available in all aws regions. From checking https://aws.amazon.com/sagemaker/pricing/ it looks like it is not available in regions: af-south-1, ap-east-1, ap-east-1, ap-southeast-3, ap-northeast-3, eu-south-1, eu-north-1, me-south-1, me-central-1 and us-gov-east-1. I believe that is the reason why sagemaker sdk doesn't specify the `instance_type`: https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/estimator.py#L1319-L1428
- What should we do around the `pins` and aws parameters mix in the function `vetiver_deploy_sagemaker`.
    - We could do something similar to the `vetiver_prepare_docker` function and have dedicated listing parameters. Maybe something like:

```r
vetiver_deploy_sagemaker <- function(board,
                                     name,
                                     instance_type,
                                     version = NULL,
                                     predict_args = list(),
                                     docker_args = list(),
                                     build_args = list(),
                                     ...) {
<<CODE>>
}
```

Reply:
OK, sounds like we should _not_ specify an instance type for this, which means [we _should_ put it up with the required arguments](https://design.tidyverse.org/args-data-details.html#args-data-details). It feels real weird to split up the pin arguments and put `instance_type` before `version`, though. Would it be weird/awkward to have it first? Or we could put `version` into a new `build_args()` argument because `vetiver_sm_build()` uses it. That does divide it up but isn't as jarring in the function reference.

Reply:
Make sense for it to be a required argument :) I think either approach would be acceptable. Probably need to consider it's api in comparison to `vetiver_deploy_rsconnect `. Should help with future deploy methods for different platforms when having that consistent-ish api template across the `vetiver_deploy_*` functions :)